pika/adapters/gevent_connection.py

```diff
@@ -0,0 +1,140 @@
+
+from pika.adapters.base_connection import BaseConnection
+from pika.heartbeat import HeartbeatChecker
+
+
+class _GeventHeartbeatChecker(HeartbeatChecker):
+    def send_and_check(self):
+        """Send and check heartbeat in a Greenlet object"""
+        gevent.spawn(super(_GeventHeartbeatChecker, self).send_and_check)
+
+
+class _GeventFakeIOLoop(object):
+    """Fake IOLoop for gevent"""
+
+    def start(self):
+        pass
+
+    def stop(self):
+        pass
+
+
+class _GeventFakeTimer(object):
+    pass
+
+
+fakeIOLoop = _GeventFakeIOLoop()
+_fakeTimer = _GeventFakeTimer()
+
+
+class GeventConnection(BaseConnection):
+    """ A standard Pika connection adapter for gevent.
+
+        :param pika.connection.Parameters parameters: Connection parameters
+        :param on_open_callback: The method to call when the connection is open
+        :type on_open_callback: method
+        :param on_open_error_callback: Method to call if the connection cant be opened
+        :type on_open_error_callback: method
+    """
+
+    def __init__(self,
+                 parameters=None,
+                 on_open_callback=None,
+                 on_open_error_callback=None,
+                 on_close_callback=None):
+        super(GeventConnection, self).__init__(
+            parameters,
+            on_open_callback,
+            on_open_error_callback,
+            on_close_callback,
+            fakeIOLoop,
+        )
+
+    def add_timeout(self, deadline, callback_method):
+        """Add the callback_method to gevent hub to fire after deadline
+        seconds.
+
+        :param int deadline: The number of seconds to wait to call callback
+        :param method callback_method: The callback method
+        :rtype: timer instance
+
+        """
+
+        if deadline:
+            timer = gevent.get_hub().loop.timer(deadline)
+            timer.start(callback_method)
+            return timer
+        else:
+            gevent.spawn(callback_method)
+            return _fakeTimer
+
+    def remove_timeout(self, timeout_id):
+        """Remove the timer from gevent hub using the instance returned from
+        add_timeout.
+
+        param: timer instance handle
+
+        """
+        timer = timeout_id
+        if timer != _fakeTimer:
+            timer.stop()
+
+    def _on_connect_timer(self):
+        gevent.spawn(super(GeventConnection, self)._on_connect_timer)
+
+    def _adapter_connect(self):
+        """Connect and read data in a Greenlet object
+
+        """
+        error = super(GeventConnection, self)._adapter_connect()
+        if not error:
+            self.socket.setblocking(1)
+            gevent.spawn(self._read_loop)
+        return error
+
+    def _adapter_disconnect(self):
+        """Disconnect the connection
+
+        """
+        super(GeventConnection, self)._adapter_disconnect()
+
+    def _handle_error(self, error_value):
+        if self.socket:
+            super(GeventConnection, self)._handle_error(error_value)
```

Comment:
Interesting, if there is an error reported to `_handle_error`, but there is no socket presently, the error is just suppressed? This doesn't sound right. If there is a legitimate reason for it, then a code comment is in order.

And again, should be checking `self.socket is not None` as objects may override `__bool__` and give it some unpredictable semantic.